hnix-store-core/src/System/Nix/Internal/Hash.hs

```diff
@@ -0,0 +1,152 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE PolyKinds           #-}
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE TypeFamilies        #-}
+{-# LANGUAGE KindSignatures      #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications    #-}
+
+module System.Nix.Internal.Hash where
+
+import qualified Crypto.Hash.MD5       as MD5
+import qualified Crypto.Hash.SHA1      as SHA1
+import qualified Crypto.Hash.SHA256    as SHA256
+import qualified Data.ByteString       as BS
+import qualified Data.ByteString.Char8 as BSC
+import           Data.Bits             (xor)
+import qualified Data.ByteString       as BS
+import qualified Data.ByteString.Lazy  as BSL
+import qualified Data.Hashable         as DataHashable
+import           Data.List             (foldl')
+import           Data.Proxy            (Proxy(Proxy))
+import qualified Data.Text             as T
+import qualified Data.Text.Encoding    as T
+import qualified Data.Vector           as V
+import           Data.Word             (Word8)
+
+-- | A tag for different hashing algorithms
+--   Also used as a type-level tag for hash digests
+--   (e.g. @Digest SHA256@ is the type for a sha256 hash)
+data HashAlgorithm
+  = MD5
+  | SHA1
+  | SHA256
+
+
+-- | Types with kind @HashAlgorithm@ may be a @HasDigest@ instance
+--   if they are able to hash bytestrings via the init/update/finalize
+--   API of cryptonite
+--
+--   Each instance defined here simply defers to one of the underlying
+--   monomorphic hashing libraries, such as `cryptohash-sha256`.
+class HasDigest (a :: HashAlgorithm) where
+
+  type AlgoCtx a :: *
+  hashLength        :: Proxy a -> Int
+
+  validateHashBytes :: BS.ByteString -> Maybe (Digest a)
```

Comment:
Hmm... Do we need this? We might, but my inclination is to not expose it until we have a use case.